### PR TITLE
Fix few problems reported by PHPStan

### PR DESCRIPTION
**Pull request by szepeviktor**
@spekulatius If you like robust code you may continue fixing the rest of the problems.

(otherwise just close it - this is an inspiration)

### CODE REVIEW

File: src/PHPScraper.php [hunk 1]
```diff
@@ -39,7 +42,7 @@ public function __construct(?array $config = [])
     /**
      * Sets the config, generates the required Clients and updates the core with the new clients.
      *
-     * @param ?array $config = []
+     * @param array<string, mixed>|null $config
```
Comment (szepeviktor) [inline src/PHPScraper.php:45]: Here we really need an _array shape_.
`array{elem1: string, elem2: bool}`
I am out of momentum.

File: phpstan.neon [hunk 2]
```diff
@@ -4,8 +4,9 @@ includes:
 parameters:
   level: 9
   paths:
-    - src
-    - tests
-  excludePaths:
-    - vendor
```
Comment (szepeviktor) [inline phpstan.neon:10]: `vendor` was no included.

File: src/UsesContent.php [hunk 3]
```diff
@@ -303,17 +303,22 @@ public function outlineWithParagraphs(): array
...
-        foreach ($result as $index => $array) {
-            if ($array[1] !== '') {
-                $result[$index] = array_combine(['tag', 'content'], $array);
-                $result[$index]['content'] = trim($result[$index]['content']);
+        /** @var array<string> $nameAndText */
+        foreach ($elementsNameAndText as $index => $nameAndText) {
+            // Element has no text.
+            if ($nameAndText[1] === '') {
+                continue;
             }
```
Comment (szepeviktor) [inline src/UsesContent.php:313]: It is beneficial to write positive condition.
```
if (error) unhappy path;

happy path;
```

[_source_](https://github.com/szepeviktor/very-clean-code)

File: src/UsesFileParsers.php [hunk 4]
```diff
@@ -279,7 +283,7 @@ public function parseJson(?string $jsonStringOrUrl = null): array
             if ($jsonStringOrUrl !== null) {
                 // Simple: Try to parse what we have been given
                 try {
-                    $result = json_decode($jsonStringOrUrl, true);
+                    $result = json_decode($jsonStringOrUrl, true, 512, JSON_THROW_ON_ERROR);
```
Comment (szepeviktor) [inline src/UsesFileParsers.php:286]: Make it _throw_!

**Comment by szepeviktor** [comment]
❌ `NavigationTest::testSurfWithAbsoluteLink` fails as link on https://test-pages.phpscraper.de/navigation/1.html changed.

**Comment by spekulatius** [comment]
Hey @szepeviktor,

Thank you for taking the time to prepare a PR! I can see there are areas that need improving. 

I've replied to your comments and changes. Could you have a look and check the PHPStan issues?

Do you think anything else should go into this?

Cheers,
Peter

**Comment by szepeviktor** [comment]
> Do you think anything else should go into this?

Yes. Your work :)

File: src/PHPScraper.php (hunk 1 above)
Comment (szepeviktor) [reply src/PHPScraper.php]: Done 🍏

File: src/PHPScraper.php [hunk 5]
```diff
@@ -11,23 +11,29 @@
...
-     * @var \Spekulatius\PHPScraper\Core;
+     * @var \Spekulatius\PHPScraper\Core
      */
     protected $core = null;
 
-    public function __construct(?array $config = [])
+    /**
+     * @param PHPScraperConfig $config
+     */
+    public function __construct(array $config = [])
```
Comment (szepeviktor) [inline src/PHPScraper.php:36]: Removed null as the base value of the configuration array.

[_source_](https://github.com/szepeviktor/very-clean-code/blob/master/README.md#-base-value-of-an-array)

File: src/PHPScraper.php [hunk 6]
```diff
@@ -39,9 +45,9 @@ public function __construct(?array $config = [])
     /**
      * Sets the config, generates the required Clients and updates the core with the new clients.
      *
-     * @param ?array $config = []
+     * @param PHPScraperConfig $config
```
Comment (szepeviktor) [inline src/PHPScraper.php:48]: These could be if your IDE complains.
```php
     * @param array $config
     * @phpstan-param PHPScraperConfig $config
```

**Comment by szepeviktor** [comment]
The next step is to remove `ignoreErrors` from PHPStan configuration.
It is up to you!

File: src/PHPScraper.php [hunk 7]
```diff
@@ -11,23 +11,29 @@
 use Symfony\Component\BrowserKit\HttpBrowser;
 use Symfony\Component\HttpClient\HttpClient as SymfonyHttpClient;
 
+/**
```
Comment (spekulatius) [inline src/PHPScraper.php:14]: Nice, that's a great solution to telling PHPStan better what is in the variables. Thanks!

File: src/PHPScraper.php (hunk 5 above)
Comment (spekulatius) [reply src/PHPScraper.php:36]: That should be fine

Comment (szepeviktor) [reply src/PHPScraper.php:36]: BTW arrays are not for carrying nonhomogeneous elements.
Arrays are for lists: keyless elements of the same type.
Configuration should be an object.
```php
$config->followRedirects();
$config->dontFollowRedirects();
```

Comment (spekulatius) [reply src/PHPScraper.php:36]: Yeah, introducing an object to contain the config would be neat

**Pull request merged**